server/rest/src/main/java/org/infinispan/rest/HeaderCorrelatorHandler.java

```diff
@@ -0,0 +1,40 @@
+package org.infinispan.rest;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http2.HttpConversionUtil;
+import io.netty.util.AsciiString;
+
+/**
+ * Handler to propagate HTTP headers between requests and responses.
+ *
+ * @since 12.0
+ */
+class HeaderCorrelatorHandler extends ChannelDuplexHandler {
+
+   public static final AsciiString STREAM_ID_HEADER = HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text();
+   private Integer streamId;
```

Comment:
maybe it is a stupid question, but isn't this instance shared between requests? I mean, this scenario:
```
req1: sets streamId=1
req2: sets streamId=2
resp from req1: read streamId==2
resp from req2: read streamId==2 again?
```

Reply:
It is only shared if annotated with ```@Sharable```. If not, it will have a new instance per pipeline

https://netty.io/4.0/api/io/netty/channel/ChannelHandler.Sharable.html

Reply:
I will double check the threading since you mentioned. Maybe a volatile is needed in the streamId.

Reply:
This is fine @pruivo. According to https://netty.io/wiki/new-and-noteworthy-in-4.0.html#wiki-h2-33:

```
There is always happens-before relationship between each ChannelHandler method invocations made by Netty. 
A user does not need to define a volatile field to keep the state of a handler.
```

and

```
Netty will never call a ChannelHandler's methods concurrently, unless the ChannelHandler is 
annotated with @Sharable.
```

But I did find an issue with this handler with HTTP/2. Since it's a multiplexing protocol (several requests identified by ```streamId``` travel through the same connection), Netty must create a child channel for each stream id to avoid the problem you describe. There was a bogus http2 pipeline that was not doing it. I removed it.